From c10e1571684a87e5974d482e13e731c69ab1797b Mon Sep 17 00:00:00 2001
From: David Daney <david.daney@cavium.com>
Date: Fri, 2 Sep 2011 11:08:46 -0700
Subject: [PATCH 136/238] netdev: octeon-ethernet: Fixes to work with MDIO multiplexers.

Source: Cavium SDK 2.2-414

If built-in we must initialize later so that the late initializing
MDIO multiplexers will have been initialized.

Remove ebb6600 hack.

Always use PHY drivers to get link status if available.

Signed-off-by: David Daney <david.daney@cavium.com>
Integrated-by: Yang Shi <yang.shi@windriver.com>
---
 drivers/net/octeon/ethernet-sgmii.c |   12 +++++++++---
 drivers/net/octeon/ethernet.c       |    4 ++++
 2 files changed, 13 insertions(+), 3 deletions(-)

diff --git a/drivers/net/octeon/ethernet-sgmii.c b/drivers/net/octeon/ethernet-sgmii.c
index 5060785..853a2d5 100644
--- a/drivers/net/octeon/ethernet-sgmii.c
+++ b/drivers/net/octeon/ethernet-sgmii.c
@@ -49,9 +49,15 @@ int cvm_oct_sgmii_open(struct net_device *dev)
 	cvmx_write_csr(CVMX_GMXX_PRTX_CFG(priv->interface_port, priv->interface), gmx_cfg.u64);
 
 	if (!octeon_is_simulation()) {
-		link_info = cvmx_helper_link_get(priv->ipd_port);
-		if (!link_info.s.link_up)
-			netif_carrier_off(dev);
+		if (priv->phydev) {
+			int r = phy_read_status(priv->phydev);
+			if (r == 0 && priv->phydev->link == 0)
+				netif_carrier_off(dev);
+		} else {
+			link_info = cvmx_helper_link_get(priv->ipd_port);
+			if (!link_info.s.link_up)
+				netif_carrier_off(dev);
+		}
 	}
 
 	return 0;
diff --git a/drivers/net/octeon/ethernet.c b/drivers/net/octeon/ethernet.c
index 286560e..383f3eb 100644
--- a/drivers/net/octeon/ethernet.c
+++ b/drivers/net/octeon/ethernet.c
@@ -1089,5 +1089,9 @@ static void __exit cvm_oct_cleanup_module(void)
 MODULE_LICENSE("GPL");
 MODULE_AUTHOR("Cavium Networks <support@caviumnetworks.com>");
 MODULE_DESCRIPTION("Cavium Networks Octeon ethernet driver.");
+#ifdef MODULE
 module_init(cvm_oct_init_module);
+#else
+late_initcall(cvm_oct_init_module);
+#endif
 module_exit(cvm_oct_cleanup_module);
-- 
1.7.0

